common-controller/internal/web/notify_handler.go

```diff
@@ -0,0 +1,218 @@
+}
+
+type jWTClaims struct {
+	Jti string `json:"jti"`
+	Exp int64  `json:"exp"`
+}
+
+var (
+	redisAddr       string
+	redisUsername   string
+	redisPassword   string
+	redisUserCertPath   string
+	redisUserKeyPath    string
+	redisCACertPath string
+	isTLSEnabled    bool
+	redisRevokedTokenChannel string
+	tokenExpiryDivider = "_##_"
+	authKeyPath string
+	authKeyHeader string
+)
+
+func init() {
+	conf := config.ReadConfigs()
+	redisHost := conf.CommonController.Redis.Host
+	redisPort := conf.CommonController.Redis.Port
+	redisAddr = redisHost + ":" + redisPort
+	redisUsername = conf.CommonController.Redis.Username
+	redisPassword = conf.CommonController.Redis.Password
+	redisUserCertPath = conf.CommonController.Redis.UserCertPath
+	redisUserKeyPath = conf.CommonController.Redis.UserKeyPath
+	redisCACertPath = conf.CommonController.Redis.CACertPath
+	isTLSEnabled = conf.CommonController.Redis.TLSEnabled
+	redisRevokedTokenChannel = conf.CommonController.Redis.RevokedTokenChannel
+	authKeyPath = conf.CommonController.Sts.AuthKeyPath
+	authKeyHeader = conf.CommonController.Sts.AuthKeyHeader
+}
+
+// NotifyHandler handles notify requests
+func NotifyHandler(c *gin.Context) {
+	_type := c.Query("type")
+	if _type == tokenRevocationType {
+		revokeToken(c)
+		return
+	}
+	c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid type"})
+	return
+}
+
+func revokeToken(c *gin.Context) {
+	if !authenticateTokenRevocationRequest(c) {
+		c.JSON(http.StatusUnauthorized, gin.H{"error": "Unauthorized request"})
+		return
+	}
+
+	var request revokeRequest
+	if err := c.ShouldBindJSON(&request); err != nil {
+		loggers.LoggerAPI.ErrorC(logging.PrintError(logging.Error3200, logging.MAJOR, "Error while parsing body: %v", err))
+		c.JSON(http.StatusBadRequest, gin.H{"error": "Error while parsing json payload"})
+		return
+	}
+  var jti string;
+	var expiry int64;
+	if request.Token != "" {
+		claims, err := extractClaimsFromJWT(request.Token)
+		if err != nil {
+			loggers.LoggerAPI.ErrorC(logging.PrintError(logging.Error3203, logging.MAJOR, "Error decoding token: %v", err))
+			c.JSON(http.StatusBadRequest, gin.H{"error": "Error decoding token"})
+			return
+		}
+		jti = claims.Jti
+		expiry = claims.Exp
+	} else {
+		jti = request.Jti
+		expiry = request.Expiry
+	}
+	if expiry <= time.Now().Unix() {
+		c.JSON(http.StatusBadRequest, gin.H{"error": "Token is already expired"})
+		return
+	}
+	err := storeTokenInRedis(jti, expiry)
+	if (err != nil) {
+		loggers.LoggerAPI.ErrorC(logging.PrintError(logging.Error3202, logging.MAJOR, "Error adding revoked tokens to redis: %v", err))
+		c.JSON(http.StatusBadRequest, gin.H{"error": "Failed to store the token in Redis cache"})
+		return
+	} 
+	c.JSON(http.StatusOK, gin.H{"message": "Token revoked successfully"})
+}
+
+func generateKey(jti string) string {
+	return fmt.Sprintf("%s:%s", "wso2:apk:revoked_token", jti)
+}
+
+func storeTokenInRedis(token string, expiry int64) error {
+	var rdb *redis.Client
```

Comment:
lets create connection once or use pooling

Reply:
I have moved the Redis client creation to the init function. This client is already supporting pool.